### PR TITLE
Adding Recommendation Indexer

### PR DESCRIPTION
**Pull request by dciborow**
Used for indexing inputs for recommendation models, both ALS and SAR.

### CODE REVIEW

**Comment by dciborow** [comment]
@mhamilton723 I'm not done with this one yet, but a NIT feedback is welcome at this point.

File: src/recommendation/src/main/scala/RecommendationIndexer.scala [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+  override def transform(dataset: Dataset[_]): DataFrame = {
+    getItemIndexModel.transform(getUserIndexModel.transform(dataset))
+  }
+
+  def this() = this(Identifiable.randomUID("RecommendationIndexerModel"))
+
+  val userIndexModel = new TransformerParam(this, "userIndexModel", "userIndexModel", { x: Transformer => true })
+
+  def setUserIndexModel(m: Model[_]): this.type = set(userIndexModel, m)
```
Comment (mhamilton723) [inline src/recommendation/src/main/scala/RecommendationIndexer.scala]: Make this the right type. If you want it to only take in a specific model nows your chance to enforce constraints

File: src/recommendation/src/main/scala/RecommendationIndexer.scala [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+  def getUserIndexModel: StringIndexerModel = $(userIndexModel).asInstanceOf[StringIndexerModel]
+
+  val itemIndexModel = new TransformerParam(this, "itemIndexModel", "itemIndexModel", { x: Transformer => true })
+
+  def setItemIndexModel(m: Model[_]): this.type = set(itemIndexModel, m)
```
Comment (mhamilton723) [inline src/recommendation/src/main/scala/RecommendationIndexer.scala]: likewise here

File: src/recommendation/src/main/scala/RecommendationIndexer.scala [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+  RecommendationIndexerBase with Wrappable {
+  override def copy(extra: ParamMap): RecommendationIndexerModel = defaultCopy(extra)
+
+  override def transform(dataset: Dataset[_]): DataFrame = {
+    getItemIndexModel.transform(getUserIndexModel.transform(dataset))
+  }
+
+  def this() = this(Identifiable.randomUID("RecommendationIndexerModel"))
+
+  val userIndexModel = new TransformerParam(this, "userIndexModel", "userIndexModel", { x: Transformer => true })
```
Comment (mhamilton723) [inline src/recommendation/src/main/scala/RecommendationIndexer.scala]: Is that check `{ x: Transformer => true }` really all ya need?

File: src/recommendation/src/main/scala/RecommendationIndexer.scala [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+  /** @group setParam */
+  def setRatingCol(value: String): this.type = set(ratingCol, value)
+
+  /** @group getParam */
+  def getRatingCol: String = $(ratingCol)
+
+  val ratingCol = new Param[String](this, "ratingCol", "Rating Col")
+
+  def transformSchema(schema: StructType): StructType = {
+    val userInputColName = $(userInputCol)
```
Comment (mhamilton723) [inline src/recommendation/src/main/scala/RecommendationIndexer.scala]: nit: use getters

File: src/recommendation/src/test/scala/RecommendationIndexerSpec.scala [hunk 5]
```diff
@@ -0,0 +1,26 @@
+// Copyright (C) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in project root for information.
+
+package com.microsoft.ml.spark
+
+import org.apache.spark.ml.util.MLReadable
+
+class RecommendationIndexerSpec extends RankingTestBase with EstimatorFuzzing[RecommendationIndexer] {
+  override def testObjects(): Seq[TestObject[RecommendationIndexer]] = {
+    List(new TestObject(recommendationIndexer, ratings))
```
Comment (mhamilton723) [inline src/recommendation/src/test/scala/RecommendationIndexerSpec.scala]: could you add a test with this and ALS so you can show its general

File: src/recommendation/src/main/scala/RecommendationIndexer.scala [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+    case _: StringIndexerModel => true
+    case _ => false
+  })
+
+  def setItemIndexModel(m: StringIndexerModel): this.type = set(itemIndexModel, m)
+
+  def getItemIndexModel: StringIndexerModel = $(itemIndexModel).asInstanceOf[StringIndexerModel]
+
+  def getUserIndex: Map[Int, String] = {
+    getUserIndexModel.asInstanceOf[StringIndexerModel]
```
Comment (mhamilton723) [inline src/recommendation/src/main/scala/RecommendationIndexer.scala]: nit: this cast int needed anymore

File: src/recommendation/src/test/scala/RecommendationIndexerSpec.scala [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+      .setRecommender(algo)
+
+    val recopipeline = new Pipeline()
+      .setStages(Array(recommendationIndexer, adapter))
+      .fit(ratings)
+
+    val output = recopipeline.transform(ratings)
+
+    val evaluator: RankingEvaluator = new RankingEvaluator()
+      .setK(5)
```
Comment (mhamilton723) [inline src/recommendation/src/test/scala/RecommendationIndexerSpec.scala]: Question: Why does the evaluator need K?

**Pull request merged**